qa/workunits/fs/maxentries/maxentries.sh

```diff
@@ -0,0 +1,169 @@
+#!/usr/bin/env bash
+
+set -ex
+
+function expect_false()
+{
+	set -x
+	if "$@"; then return 1; else return 0; fi
+}
+
+function make_files()
+{
+  set +x
+  temp_dir=`mktemp -d mkfile_test_XXXXXX`
+  for i in $(seq 1 $1)
+  do
+    touch ${temp_dir}/file_${i}
+    if [ $? -ne 0 ] ; then
+      return 1
+    fi
+  done
+  set -x
+  return 0
+}
+
+function make_dirs()
+{
+  set +x
+  temp_dir=`mktemp -d mkdir_test_XXXXXX`
+  for i in $(seq 1 $1)
+  do
+    mkdir -p ${temp_dir}/dir_${i}
+    if [ $? -ne 0 ] ; then
+      return 1
+    fi
+  done
+  set -x
+  return 0
+}
+
+function make_nodes()
+{
+  set +x
+  temp_dir=`mktemp -d mknod_test_XXXXXX`
+  for i in $(seq 1 $1)
+  do
+    mknod ${temp_dir}/fifo_${i} p
+    if [ $? -ne 0 ] ; then
+      return 1
+    fi
+  done
+  set -x
+  return 0
+}
+
+function rename_files()
+{
+  set +x
+  temp_dir=`mktemp -d rename_test_XXXXXX`
+  mkdir -p ${temp_dir}/rename
+
+  for i in $(seq 1 $1)
+  do
+    touch ${temp_dir}/file_${i}
+    if [ $? -ne 0 ] ; then
+      return 1
+    fi
+    mv ${temp_dir}/file_${i} ${temp_dir}/rename/
+    if [ $? -ne 0 ] ; then
+      return 1
+    fi
+  done
+  set -x
+  return 0
+}
+
+function make_symlinks()
+{
+  set +x
+  temp_dir=`mktemp -d symlink_test_XXXXXX`
+  mkdir -p ${temp_dir}/symlink 
+
+  touch ${temp_dir}/file
+
+  for i in $(seq 1 $1)
+  do
+    ln -s ../file ${temp_dir}/symlink/sym_${i}
+    if [ $? -ne 0 ] ; then
+      return 1
+    fi
+  done
+  set -x
+  return 0
+}
+
+function make_links()
+{
+  set +x
+  temp_dir=`mktemp -d link_test_XXXXXX`
+  mkdir -p ${temp_dir}/link 
+
+  touch ${temp_dir}/file
+
+  for i in $(seq 1 $1)
+  do
+    ln ${temp_dir}/file ${temp_dir}/link/link_${i}
+    if [ $? -ne 0 ] ; then
+      return 1
+    fi
+  done
+  set -x
+  return 0
+}
+
+function cleanup()
+{
+  rm -rf *
+}
+
+test_dir="max_entries"
+mkdir -p $test_dir
+pushd $test_dir
+
+dir_max_entries=100
+ceph config set mds mds_dir_max_entries $dir_max_entries
+
+ok_dir_max_entries=$dir_max_entries
+fail_dir_max_entries=$((dir_max_entries+1))
+
+# make files test
+make_files $ok_dir_max_entries
+expect_false make_files $fail_dir_max_entries
```

Comment:
````
2021-10-15T03:46:05.474 INFO:tasks.workunit.client.0.smithi090.stderr:+ expect_false make_files 101
2021-10-15T03:46:05.474 INFO:tasks.workunit.client.0.smithi090.stderr:+ set -x
2021-10-15T03:46:05.474 INFO:tasks.workunit.client.0.smithi090.stderr:+ make_files 101
2021-10-15T03:46:05.475 INFO:tasks.workunit.client.0.smithi090.stderr:+ set +x
2021-10-15T03:46:05.619 DEBUG:teuthology.orchestra.run:got remote process result: 1
2021-10-15T03:46:05.620 INFO:tasks.workunit.client.0.smithi090.stderr:+ return 0
2021-10-15T03:46:05.620 INFO:tasks.workunit.client.0.smithi090.stderr:+ return 1
2021-10-15T03:46:05.620 INFO:tasks.workunit:Stopping ['fs/maxentries'] on client.0...
2021-10-15T03:46:05.621 DEBUG:teuthology.orchestra.run.smithi090:> sudo rm -rf -- /home/ubuntu/cephtest/workunits.list.client.0 /home/ubuntu/cephtest/clone.client.0
2021-10-15T03:46:06.321 ERROR:teuthology.run_tasks:Saw exception from tasks.
Traceback (most recent call last):
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_423bd94291582cb44cbfac84a53ec84580eb3f08/teuthology/run_tasks.py", line 91, in run_tasks
    manager = run_one_task(taskname, ctx=ctx, config=config)
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_423bd94291582cb44cbfac84a53ec84580eb3f08/teuthology/run_tasks.py", line 70, in run_one_task
    return task(**kwargs)
  File "/home/teuthworker/src/git.ceph.com_ceph-c_9bcded95726562b598e8b1357a01cf828b42f563/qa/tasks/workunit.py", line 134, in task
    coverage_and_limits=not config.get('no_coverage_and_limits', None))
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_423bd94291582cb44cbfac84a53ec84580eb3f08/teuthology/parallel.py", line 84, in __exit__
    for result in self:
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_423bd94291582cb44cbfac84a53ec84580eb3f08/teuthology/parallel.py", line 98, in __next__
    resurrect_traceback(result)
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_423bd94291582cb44cbfac84a53ec84580eb3f08/teuthology/parallel.py", line 30, in resurrect_traceback
    raise exc.exc_info[1]
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_423bd94291582cb44cbfac84a53ec84580eb3f08/teuthology/parallel.py", line 23, in capture_traceback
    return func(*args, **kwargs)
  File "/home/teuthworker/src/git.ceph.com_ceph-c_9bcded95726562b598e8b1357a01cf828b42f563/qa/tasks/workunit.py", line 426, in _run_tests
    label="workunit test {workunit}".format(workunit=workunit)
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_423bd94291582cb44cbfac84a53ec84580eb3f08/teuthology/orchestra/remote.py", line 509, in run
    r = self._runner(client=self.ssh, name=self.shortname, **kwargs)
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_423bd94291582cb44cbfac84a53ec84580eb3f08/teuthology/orchestra/run.py", line 455, in run
    r.wait()
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_423bd94291582cb44cbfac84a53ec84580eb3f08/teuthology/orchestra/run.py", line 161, in wait
    self._raise_for_status()
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_423bd94291582cb44cbfac84a53ec84580eb3f08/teuthology/orchestra/run.py", line 183, in _raise_for_status
    node=self.hostname, label=self.label
teuthology.exceptions.CommandFailedError: Command failed (workunit test fs/maxentries/maxentries.sh) on smithi090 with status 1: 'mkdir -p -- /home/ubuntu/cephtest/mnt.0/client.0/tmp && cd -- /home/ubuntu/cephtest/mnt.0/client.0/tmp && CEPH_CLI_TEST_DUP_COMMAND=1 CEPH_REF=9bcded95726562b598e8b1357a01cf828b42f563 TESTDIR="/home/ubuntu/cephtest" CEPH_ARGS="--cluster ceph" CEPH_ID="0" PATH=$PATH:/usr/sbin CEPH_BASE=/home/ubuntu/cephtest/clone.client.0 CEPH_ROOT=/home/ubuntu/cephtest/clone.client.0 CEPH_MNT=/home/ubuntu/cephtest/mnt.0 adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 3h /home/ubuntu/cephtest/clone.client.0/qa/workunits/fs/maxentries/maxentries.sh'
````

From: /ceph/teuthology-archive/pdonnell-2021-10-15_02:49:27-fs-wip-pdonnell-testing-20211012.192211-distro-basic-smithi/6443978/teuthology.log

This fails with async creates on the latest kernels. You need to do a `sync $dir` to get an ENOSPC.

Reply:
It seems metadata inconsistent scenario that metadata updates are cached in the kernel and early acknowledged to the user process and then later flushed to metadata servers. Based on this assumption, a reply message with ENOSPC may be missed in the kernel.  Could you let me know there is an efficient way to create files synchronously? Are there some example cases related to this issue in qa suites? I think touching file followed by sync command is not feasible for kernel client. The reason is that creating the 101th and 10Nth files may be successfully done regardless of sync command in a short period of time. 

How about using ceph-fuse for this test to avoid this early completion issue?

Could you share the latest kernel version used in your experiments?

Reply:
> Based on this assumption, a reply message with ENOSPC may be missed in the kernel.

It wont' be missed. The kernel eventually gets an ENOSPC from the create RPC. It has always been the application's responsibility to check whether a file is durable via `fsync` on the inode and (for new files/links) an `fsync` on the containing directory inode.

>  Could you let me know there is an efficient way to create files synchronously? 

You don't need to create the files synchronously. You just need to fsync the directory before letting the shell script exit.

> he reason is that creating the 101th and 10Nth files may be successfully done regardless of sync command in a short period of time.

The `sync` command will fail instead of `touch`. That should be sufficient, no?

> How about using ceph-fuse for this test to avoid this early completion issue?

All functional tests should work regardless of client.

> Could you share the latest kernel version used in your experiments?

https://github.com/ceph/ceph-client/tree/testing